rust/lance/src/io/exec/anti_join_limit_pushdown.rs

```diff
@@ -0,0 +1,249 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright The Lance Authors
+
+//! Anti-join limit pushdown optimization
+//!
+//! This optimizer pushes down scan limits for anti-join patterns to reduce
+//! the amount of data read from the probe side table.
+//!
+//! Pattern: Limit -> HashJoin(Anti) -> LanceScan (probe side)
+//!
+//! For anti-joins with limits, we only need to scan at most:
+//! limit + build_side_rows from the probe side (worst case).
+
+use std::sync::Arc;
+
+use datafusion::common::Result as DFResult;
+use datafusion::config::ConfigOptions;
+use datafusion::logical_expr::JoinType;
+use datafusion::physical_optimizer::PhysicalOptimizerRule;
+use datafusion::physical_plan::joins::HashJoinExec;
+use datafusion::physical_plan::limit::{GlobalLimitExec, LocalLimitExec};
+use datafusion::physical_plan::ExecutionPlan;
+
+use crate::io::exec::scan::LanceScanExec;
+
+/// Helper function to get exact row count from a plan, traversing through wrapper nodes
+fn get_exact_row_count(plan: Arc<dyn ExecutionPlan>) -> Option<usize> {
+    use datafusion::common::stats::Precision;
+
+    // Try partition_statistics first (newer API), then fall back to statistics()
+    // LanceScanExec doesn't implement partition_statistics but does implement statistics()
+    let stats = match plan.partition_statistics(None) {
+        Ok(stats) => stats,
+        Err(_) => {
+            // Fall back to statistics() for nodes that don't support partition_statistics
+            match plan.statistics() {
+                Ok(stats) => stats,
+                Err(_) => return None,
+            }
+        }
+    };
+
+    if let Precision::Exact(n) = stats.num_rows {
+        return Some(n);
+    }
+
+    // If no exact statistics, try children (for wrapper nodes like CoalesceBatchesExec)
+    let children = plan.children();
+    if children.len() == 1 {
+        return get_exact_row_count(children[0].clone());
+    }
+
+    None
+}
+
+/// Optimizer that pushes down limits to LanceScan for anti-join patterns
+///
+/// This reduces I/O and memory usage by limiting how much data we read
+/// from the probe side when we know we only need a limited number of results.
+#[derive(Debug, Clone, Default)]
+pub struct AntiJoinLimitPushdown {}
+
+impl AntiJoinLimitPushdown {
+    pub fn new() -> Self {
+        Self {}
+    }
+
+    /// Push down limit to scan by using with_fetch or adding LocalLimitExec
+    fn push_limit_to_scan(
+        &self,
+        probe_side: Arc<dyn ExecutionPlan>,
+        limit: usize,
+        build_rows: usize,
+    ) -> Arc<dyn ExecutionPlan> {
+        // Calculate maximum rows we need to scan
+        // Worst case: first build_rows all match (excluded), then next limit rows don't match
+        let max_probe_rows = limit + build_rows;
```

Comment:
@LuQQiu I don't quite follow this. The comment seems incorrect to me - shouldn't the worst case be scanning all rows in each table? That would happen if there weren't enough matches to fill the limit. I don't see how the maximum probe rows would be dependent on the user's limit.

Reply:
You raised a good point. Is it based on the assumption that the values are unique. if it's not unique, the case would be very different.

Reply:
Would that hopefully mean that we can just use DataFusion's operator?